### PR TITLE
Add optional UpstreamAuthority

### PR DESCRIPTION
**Pull request by KenHuffmanAtNice**
This adds the ability to add a "disk" UpstreamAuthority

Also specify the name of a CA secret that can be used with the "disk" UpstreamAuthority

Signed-off-by: Ken Huffman <ken.huffman@niceincontact.com>

### CODE REVIEW

**Review by joe-carmo — APPROVED** [review]

File: charts/spire/Chart.yaml [hunk 1]
```diff
@@ -27,7 +27,7 @@ description: |
           - --service-account-signing-key-file=/run/config/pki/sa.key
   ```
 type: application
-version: 0.7.1
+version: 0.7.2
```
Comment (marcofranssen) [inline charts/spire/Chart.yaml:30]:
```suggestion
version: 0.8.0
```

As this brings a new feature.

File: charts/spire/templates/agent-daemonset.yaml [hunk 2]
```diff
@@ -13,6 +13,10 @@ spec:
     type: RollingUpdate
   template:
     metadata:
+      {{- with .Values.agent.service.annotations }}
```
Comment (marcofranssen) [inline charts/spire/templates/agent-daemonset.yaml:16]: Could we do this change as a separate PR to patch the 0.7.x version?

File: charts/spire/templates/agent-daemonset.yaml [hunk 3]
```diff
@@ -28,6 +32,8 @@ spec:
           image: {{ template "spire.image" .Values.waitForIt }}
           imagePullPolicy: {{ .Values.waitForIt.image.pullPolicy }}
           args: ["-t", "30", "{{ include "spire.fullname" . }}-server:8081"]
+          resources:
```
Comment (marcofranssen) [inline charts/spire/templates/agent-daemonset.yaml:35]: Could we do this change as a separate PR to patch the 0.7.x version?

File: charts/spire/templates/agent-daemonset.yaml [hunk 4]
```diff
@@ -45,7 +51,7 @@ spec:
               mountPath: /run/spire/bundle
               readOnly: true
             - name: spire-agent-socket-dir
-              mountPath: /run/spire/agent-sockets
+              mountPath: {{ dir .Values.agent.config.socketPath }}
```
Comment (marcofranssen) [inline charts/spire/templates/agent-daemonset.yaml:54]: Could we do this change as a separate PR to patch the 0.7.x version?

File: charts/spire/templates/agent-daemonset.yaml [hunk 5]
```diff
@@ -105,7 +111,17 @@ spec:
           args: [
             "-csi-address", "/spiffe-csi/csi.sock",
             "-kubelet-registration-path", "/var/lib/kubelet/plugins/csi.spiffe.io/csi.sock",
+            "-http-endpoint", ":9809",
```
Comment (marcofranssen) [inline charts/spire/templates/agent-daemonset.yaml:114]: Could you explain this feature? Also maybe isolate that in separate PR.

File: charts/spire/templates/agent-daemonset.yaml [hunk 6]
```diff
@@ -125,7 +141,7 @@ spec:
             name: {{ include "spire.fullname" . }}-bundle
         - name: spire-agent-sockets
           hostPath:
-            path: /run/spire/agent-sockets
+            path: {{ dir .Values.agent.config.socketPath }}
```
Comment (marcofranssen) [inline charts/spire/templates/agent-daemonset.yaml:144]: Could we do this change as a separate PR to patch the 0.7.x version?

File: charts/spire/templates/server-service.yaml [hunk 7]
```diff
@@ -3,6 +3,10 @@ kind: Service
 metadata:
   name: {{ include "spire.fullname" . }}-server
   namespace: {{ .Release.Namespace }}
+  {{- with .Values.server.service.annotations }}
```
Comment (marcofranssen) [inline charts/spire/templates/server-service.yaml:6]: Could we do this change as a separate PR to patch the 0.7.x version?

File: charts/spire/values.yaml [hunk 8]
```diff
@@ -63,6 +67,7 @@ server:
   service:
     type: ClusterIP
     port: 8081
+    annotations: {}
```
Comment (marcofranssen) [inline charts/spire/values.yaml:70]: Could we do this change as a separate PR to patch the 0.7.x version?

File: charts/spire/values.yaml [hunk 9]
```diff
@@ -75,6 +80,13 @@ server:
     # runAsNonRoot: true
     # runAsUser: 1000
 
+  topologySpreadConstraints: []
```
Comment (marcofranssen) [inline charts/spire/values.yaml:83]: Could we do this change as a separate PR to patch the 0.7.x version?

File: charts/spire/values.yaml [hunk 10]
```diff
@@ -85,6 +97,20 @@ server:
       organization: Example
       common_name: example.org
 
+    upstreamAuthority:
+      certManager:
+        enabled: false
+        issuerName: "spire-ca"
```
Comment (marcofranssen) [inline charts/spire/values.yaml:103]:
```suggestion
        issuerName: spire-ca
```

File: charts/spire/values.yaml [hunk 11]
```diff
@@ -85,6 +97,20 @@ server:
...
+        issuerKind: "Issuer"
```
Comment (marcofranssen) [inline charts/spire/values.yaml:104]:
```suggestion
        issuerKind: Issuer
```

File: charts/spire/values.yaml [hunk 12]
```diff
@@ -85,6 +97,20 @@ server:
...
+        issuerGroup: "cert-manager.io"
```
Comment (marcofranssen) [inline charts/spire/values.yaml:105]:
```suggestion
        issuerGroup: cert-manager.io
```

File: charts/spire/values.yaml [hunk 13]
```diff
@@ -85,6 +97,20 @@ server:
...
+        namespace: "sandbox"
```
Comment (marcofranssen) [inline charts/spire/values.yaml:106]:
```suggestion
        namespace: cert-manager
```

I assume most people will by default deploy to this namespace.

File: charts/spire/values.yaml [hunk 14]
```diff
@@ -85,6 +97,20 @@ server:
...
+        kubeConfigFile: ""
+      disk:
+        enabled: false
+        certFilePath: "/conf/server/tls.crt"
```
Comment (marcofranssen) [inline charts/spire/values.yaml:110]:
```suggestion
        certFilePath: /conf/server/tls.crt
```

File: charts/spire/values.yaml [hunk 15]
```diff
@@ -85,6 +97,20 @@ server:
...
+        keyFilePath: "/conf/server/tls.key"
```
Comment (marcofranssen) [inline charts/spire/values.yaml:111]:
```suggestion
        keyFilePath: /conf/server/tls.key
```

File: charts/spire/values.yaml [hunk 16]
```diff
@@ -110,6 +136,9 @@ agent:
     #   cpu: 100m
     #   memory: 128Mi
 
+  service:
```
Comment (marcofranssen) [inline charts/spire/values.yaml:139]: Could we do this change as a separate PR to patch the 0.7.x version?

**Review by marcofranssen — CHANGES_REQUESTED** [review]
🎉  Thanks a lot for this great contribution.

In our CONTRIBUTING.md we described we prefer to have one feature / improvement per PR to reduce the risk and scope for us, therefore a few requests below.

I did a few style suggestions in the values file so it is aligned with style used in the rest of the values.yaml.

Furthermore I would love to take all the other small improvements as a patch release (could you move those commits into a new branch/pr?).

Regarding cert-manager.io I have a couple of questions and would love to understand how to use the feature so I can also do some testing on our own infrastructure.

File: charts/spire/templates/agent-daemonset.yaml (hunk 2 above)
Comment (KenHuffmanAtNice) [reply charts/spire/templates/agent-daemonset.yaml:16]: Makes sense, are you okay will all the non-UpstreamAuthority related changes being in same patch version?

File: charts/spire/templates/agent-daemonset.yaml (hunk 5 above)
Comment (KenHuffmanAtNice) [reply charts/spire/templates/agent-daemonset.yaml:114]: This flag adds a livenessProbe to the node-driver-registrar and goes with livenessProbe definition I added a few lines later. My company likes to turn on liveness probes whenever possible, but this change is entirely optional.

File: charts/spire/templates/server-statefulset.yaml [hunk 17]
```diff
@@ -116,6 +127,11 @@ spec:
           hostPath:
             path: /run/spire/server-sockets
             type: DirectoryOrCreate
+        {{- if eq (.Values.server.caSecretStorage.enabled | toString) "true" }}
```
[the remark this answers is not in the document]
Comment (KenHuffmanAtNice) [reply charts/spire/templates/server-statefulset.yaml:130]: See my previous answer :-)

File: charts/spire/README.md [hunk 18]
```diff
@@ -30,6 +31,7 @@ spec:
         - --service-account-signing-key-file=/run/config/pki/sa.key
 ```
 
+
```
Comment (marcofranssen) [inline charts/spire/README.md:34]: Seems you have an issue with line endings on Windows using helm-docs.

Comment (KenHuffmanAtNice) [reply charts/spire/README.md:34]: These unusual whitespace changes were made by helm_docs.sh

**Review by marcofranssen — APPROVED** [review]

**Review by marcofranssen — CHANGES_REQUESTED** [review]
The docs are still failing, please allow me to fix it on this branch or commit the fix. Otherwise fine to merge this PR.

I have an upcomping refactor coming in, would love to merge this one first to not disturb this PR with merge conflicts.

@KenHuffmanAtNice

**Comment by KenHuffmanAtNice** [comment]
> The docs are still failing, please allow me to fix it on this branch or commit the fix.

@marcofranssen Added you as collaborator.

**Comment by marcofranssen** [comment]
> > The docs are still failing, please allow me to fix it on this branch or commit the fix.
> 
> @marcofranssen Added you as collaborator.

Just check this checkbox on the PR itself. That should allow me to push.

image

**Review by marcofranssen — APPROVED** [review]
